opm/autodiff/SimulatorFullyImplicitBlackoilOutputEbos.cpp

```diff
@@ -0,0 +1,282 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "config.h"
+
+#include "SimulatorFullyImplicitBlackoilOutputEbos.hpp"
+
+#include <opm/common/data/SimulationDataContainer.hpp>
+
+#include <opm/parser/eclipse/EclipseState/InitConfig/InitConfig.hpp>
+#include <opm/core/simulator/BlackoilState.hpp>
+#include <opm/core/utility/DataMap.hpp>
+#include <opm/autodiff/Compat.hpp>
+#include <opm/output/vtk/writeVtkData.hpp>
+#include <opm/common/ErrorMacros.hpp>
+#include <opm/core/utility/miscUtilities.hpp>
+#include <opm/parser/eclipse/Units/Units.hpp>
+
+#include <opm/autodiff/GridHelpers.hpp>
+#include <opm/autodiff/BackupRestore.hpp>
+
+#include <sstream>
+#include <iomanip>
+#include <fstream>
+
+#include <boost/filesystem.hpp>
+
+//For OutputWriterHelper
+#include <map>
+#include <opm/parser/eclipse/Units/UnitSystem.hpp>
+
+
+#ifdef HAVE_OPM_GRID
+#include <opm/common/utility/platform_dependent/disable_warnings.h>
+#include <dune/common/version.hh>
+#include <dune/grid/io/file/vtk/vtkwriter.hh>
+#include <opm/common/utility/platform_dependent/reenable_warnings.h>
+#endif
+namespace Opm
+{
+    namespace detail {
+
+        struct WriterCallEbos : public ThreadHandle :: ObjectInterface
+        {
+            BlackoilOutputWriterEbos& writer_;
+            std::unique_ptr< SimulatorTimerInterface > timer_;
+            const SimulationDataContainer state_;
+            const WellState wellState_;
+            data::Solution simProps_;
+            const bool substep_;
+
+            explicit WriterCallEbos( BlackoilOutputWriterEbos& writer,
+                                 const SimulatorTimerInterface& timer,
+                                 const SimulationDataContainer& state,
+                                 const WellState& wellState,
+                                 const data::Solution& simProps,
+                                 bool substep )
+                : writer_( writer ),
+                  timer_( timer.clone() ),
+                  state_( state ),
+                  wellState_( wellState ),
+                  simProps_( simProps ),
+                  substep_( substep )
+            {
+            }
+
+            // callback to writer's serial writeTimeStep method
+            void run ()
+            {
+                // write data
+                writer_.writeTimeStepSerial( *timer_, state_, wellState_, simProps_, substep_ );
+            }
+        };
+    }
+
+    void
+    BlackoilOutputWriterEbos::
+    writeTimeStepWithoutCellProperties(
+                  const SimulatorTimerInterface& timer,
+                  const SimulationDataContainer& localState,
+                  const WellState& localWellState,
+                  bool substep)
+    {
+        data::Solution noCellProperties;
```

Comment:
This class (this method for example, but also others) must be rewritten to work properly in parallel. We have changed the order so that the "simToSolution()" method extracting the "common" data like pressure to the data::Solution object is called first, then the asked-for data (from summary and restart configs) are added, then (in WriteTimeStepWithCellProperties() as before) the data is gathered (but now also gathering correctly the data::Solution object).